modules/elasticsearch/src/main/java/org/testcontainers/elasticsearch/ElasticsearchContainer.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to David Pilato (the "Author") under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Author licenses this
+ * file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.testcontainers.elasticsearch;
+
+import org.apache.http.HttpHost;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+
+import java.io.IOException;
+import java.util.Properties;
+
+/**
+ * Represents an elasticsearch docker instance which exposes by default port 9200 and 9300 (transport.tcp.port)
+ * The docker image is by default fetch from docker.elastic.co/elasticsearch/elasticsearch
+ * @author dadoonet
+ */
+public class ElasticsearchContainer<SELF extends ElasticsearchContainer<SELF>> extends GenericContainer<SELF> {
+
+    private static final String FALLBACK_RESOURCE_NAME = "elasticsearch-default.properties";
+    private static final int ELASTICSEARCH_DEFAULT_PORT = 9200;
+    private static final int ELASTICSEARCH_DEFAULT_TCP_PORT = 9300;
+    static final String ELASTICSEARCH_DEFAULT_BASE_URL;
+    static final String ELASTICSEARCH_DEFAULT_VERSION;
+    static {
+        Properties props = new Properties();
+        try {
+            props.load(ElasticsearchContainer.class.getResourceAsStream(ElasticsearchContainer.FALLBACK_RESOURCE_NAME));
+        } catch (IOException ignored) {
+        }
+        ELASTICSEARCH_DEFAULT_BASE_URL = props.getProperty("baseUrl");
+        ELASTICSEARCH_DEFAULT_VERSION = props.getProperty("version");
+    }
+
+    private String baseUrl = ELASTICSEARCH_DEFAULT_BASE_URL;
+    private String version = ELASTICSEARCH_DEFAULT_VERSION;
+
+    /**
+     * Define the elasticsearch version to start
+     * @param version  Elasticsearch Version like 5.6.10 or 6.3.2
+     * @return this
+     */
+    public ElasticsearchContainer withVersion(String version) {
+        this.version = version;
+        return this;
+    }
+
+    /**
+     * Define the elasticsearch docker registry base url
+     * @param baseUrl  defaults to docker.elastic.co/elasticsearch/elasticsearch
+     * @return this
+     */
+    public ElasticsearchContainer withBaseUrl(String baseUrl) {
```

Comment:
Using the `withBaseUrl()` and `withVersion()` builder methods is an interesting approach. Do you have a strong reason for this? The "idiomatic" Testcontainers way would be, to define those constructor parameters (the others can correct me here, if I'm mistaken).

Reply:
I wanted to provide OOTB values. So people would just have to write:

```java
new ElasticsearchContainer();
```

Instead of:

```java
new ElasticsearchContainer("6.4.1");
```

That's even easier if you have multiple options you want to set.
Instead of providing multiple CTORs, I can just have the default one plus builders to change parameters.

```java
new ElasticsearchContainer("6.4.1");
new ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch", "6.4.1");
```

That's not super obvious in that case but if you think of other options I'd like to add later on, I think this way brings more flexibility.

Reply:
Usually, we have two options:
```
new ElasticsearchContainer()
```

and
```
new ElasticsearchContainer("my.corp.org/registry/elasticsearch:fancy_version")
```

One of the reasons to have it (and only it) as a constructor parameter is that we can fail fast if image is not available (subject to at least discuss but for now it is so)

Reply:
I added also this option. So we have both (using `.withXYZ()` and CTOR)

Reply:
constructors should call super, please check other modules for the ref
Also, I would really avoid `withXYZ` for the image. 
It is at least inconsistent with the rest of our modules... There is also a naming issue, it is not a `baseUrl`, but `repository`, in Docker terms

Reply:
I disagree because as a end user, I'd definitely prefer writing:

```java
new ElasticsearchContainer().withVersion("6.4.1").start();
```

Than:

```java
new ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch:6.4.1").start();
```

But I'll follow your pattern.

Reply:
We had some plans how we can rework the DSL (also see #658 ), but for now I would keep the consistency between the modules. However, it would be nice to get your feedback on a new DSL when we will be actively working on it 👍